src/fast_graph32.rs

```diff
@@ -0,0 +1,223 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+use std::convert::TryFrom;
+
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::fast_graph::FastGraphEdge;
+use crate::FastGraph;
+
+/// Special graph data-structure that is identical to `FastGraph` except that it uses u32 integers
+/// instead of usize integers. This is used to store a `FastGraph` in a 32bit representation on disk
+/// when using a 64bit system.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct FastGraph32 {
+    num_nodes: u32,
+    pub ranks: Vec<u32>,
+    pub edges_fwd: Vec<FastGraphEdge32>,
+    pub first_edge_ids_fwd: Vec<u32>,
+
+    pub edges_bwd: Vec<FastGraphEdge32>,
+    pub first_edge_ids_bwd: Vec<u32>,
+}
+
+impl FastGraph32 {
+    /// Creates a 32bit Graph from a given `FastGraph`. All (potentially 64bit) `usize` integers are
+    /// simply converted to u32 and if a value exceeds the 32bit limit an error is thrown. The only
+    /// exception is `std::u32::MAX`, which is converted to `std::usize::MAX`.
+    pub fn new(fast_graph: &FastGraph) -> Self {
+        FastGraph32 {
+            num_nodes: usize_to_u32(fast_graph.get_num_nodes()),
+            ranks: usize_to_u32_vec(&fast_graph.ranks),
+            edges_fwd: usize_to_u32_edges(&fast_graph.edges_fwd),
+            first_edge_ids_fwd: usize_to_u32_vec(&fast_graph.first_edge_ids_fwd),
+            edges_bwd: usize_to_u32_edges(&fast_graph.edges_bwd),
+            first_edge_ids_bwd: usize_to_u32_vec(&fast_graph.first_edge_ids_bwd),
+        }
+    }
+
+    /// Converts a 32bit Graph to an actual `FastGraph` using `usize` such that it can be used with
+    /// FastPaths crate. Any integers that equal `std::u32::MAX` are mapped to `std::usize::MAX`.
+    pub fn convert_to_usize(self) -> FastGraph {
+        let mut g = FastGraph::new(self.num_nodes as usize);
+        g.ranks = u32_to_usize_vec(&self.ranks);
+        g.edges_fwd = u32_to_usize_edges(&self.edges_fwd);
+        g.first_edge_ids_fwd = u32_to_usize_vec(&self.first_edge_ids_fwd);
+        g.edges_bwd = u32_to_usize_edges(&self.edges_bwd);
+        g.first_edge_ids_bwd = u32_to_usize_vec(&self.first_edge_ids_bwd);
+        g
+    }
+}
+
+/// 32bit equivalent to `FastGraphEdge`, see `FastGraph32` docs.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct FastGraphEdge32 {
+    pub base_node: u32,
+    pub adj_node: u32,
+    pub weight: u32,
+    pub replaced_in_edge: u32,
+    pub replaced_out_edge: u32,
+}
+
+fn usize_to_u32(int: usize) -> u32 {
+    if int == std::usize::MAX {
+        usize_to_u32(std::u32::MAX as usize)
+    } else {
+        if let Ok(x) = u32::try_from(int) {
+            x
+        } else {
+            panic!("Could not convert {} to a 32-bit integer", int);
+        }
+    }
+}
+
+fn usize_to_u32_vec(vec: &Vec<usize>) -> Vec<u32> {
+    vec.iter().map(|i| usize_to_u32(*i)).collect()
```

Comment:
@dabreegster do you know how this can be changed such that while we are traversing vec we are releasing the memory (and only keep the transformed vector here?). doing it this way we should be able to remove any memory overhead when creating FG32

Reply:
I don't think it's possible for serialization. The [API](https://docs.serde.rs/serde/trait.Serialize.html) borrows the object to serialize; it doesn't consume it. I couldn't find anyone discussing a zero-copy serde serialization, but it could be worth searching the project's issues or asking around.

Deserialization on the other hand is zero-copy: https://serde.rs/lifetimes.html

Reply:
Hm, but why is this related to serde? I thought we could change it such that when creating FG32 we let go of the original graph. Or do you mean if we do this this can no longer be used with serde?

Reply:
This would force somebody to clone() when using with serde. And why would somebody convert to FG32 if they're not serializing?

Reply:
Yes FG32 is only meant for serializing and if serde imposes this restriction that we cannot consume I do not have a better idea how to do this either.